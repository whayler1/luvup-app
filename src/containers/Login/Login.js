import React, { Component } from 'react';
import superagent from 'superagent';
import { Actions } from 'react-native-router-flux';

import config from '../../config.js';
import Template from './Login.template';


export default class Login extends Component {
  state = {
    username: '',
    password: '',
    error: '',
  }

  onUsernameChange = username => this.setState({ username });
  onPasswordChange = password => this.setState({ password });

  getValidationError = () => {
    const { username, password } = this.state;
    if (!username) {
      return 'username';
    }
    if (!password) {
      return 'password';
    }
    return '';
  }

  navigateToSignUp = () => Actions.signup();
  navigateToSignUpConfirm = () => Actions.signupconfirm();

  onSubmit = async () => {
    const errorStr = this.getValidationError();
    this.setState({ error: errorStr });
    if (errorStr) {
      return;
    }
    const { username, password } = this.state;
    try {
      const res = await superagent.post(`${config.baseUrl}/login`, {
        username,
        password
      });

      console.log('Login res!', res.body);
    } catch(err) {
      this.setState({ error: 'server' });
      console.log('error', err);
    }
  }

  render() {
    const { username, password, error } = this.state;

    return <Template
      navigateToSignUpConfirm={this.navigateToSignUpConfirm}
      navigateToSignUp={this.navigateToSignUp}
      onSubmit={this.onSubmit}
      onUsernameChange={this.onUsernameChange}
      onPasswordChange={this.onPasswordChange}
      username={username}
      password={password}
      error={error}
    />;
  }
};
