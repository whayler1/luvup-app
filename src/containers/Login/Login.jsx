import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Actions } from 'react-native-router-flux';
import { View, Text, TouchableOpacity } from 'react-native';

import styles from './Login.styles';
import { scene, forms } from '../../styles';
import { login as loginAction } from '../../redux/user/user.actions';
import Input from '../../components/Input';
import Well from '../../components/Well';
import FormScene from '../../components/FormScene';
import Button, { BUTTON_STYLES } from '../../components/Button';

let passwordInput;
const focusPassword = () => passwordInput.focus();

class Login extends Component {
  static propTypes = {
    username: PropTypes.string,
    login: PropTypes.func.isRequired,
    getMeErrorMessage: PropTypes.string.isRequired,
    isLoginInFlight: PropTypes.bool.isRequired,
    loginError: PropTypes.string.isRequired,
  };

  static defaultProps = {
    username: '',
  };

  constructor(props) {
    super(props);

    this.state = {
      username: props.username,
      password: '',
      error: '',
    };
  }

  handleUsernameChange = (username) => {
    this.setState({ username });
  };
  handlePasswordChange = (password) => {
    this.setState({ password });
  };

  getValidationError = () => {
    const { username, password } = this.state;
    if (!username) {
      return 'username';
    }
    if (!password) {
      return 'password';
    }
    return '';
  };

  getFormattedLoginError() {
    const { loginError } = this.props;
    if (loginError === 'Unsuccessful HTTP response') {
      return 'Invalid email or password';
    }
    return loginError;
  }

  navigateToSignUp = () => Actions.signup();
  navigateToSignUpConfirm = () => Actions.confirmUserRequestCode();
  navigateToForgotPassword = () => Actions.forgotPassword();

  submit = async () => {
    const { username, password } = this.state;
    this.props.login(username, password);
  };

  handleSubmit = () => {
    const errorStr = this.getValidationError();
    this.setState({ error: errorStr });
    if (errorStr) {
      return;
    }
    this.submit();
  };

  setPasswordInputRef = (el) => {
    passwordInput = el;
  };

  render() {
    const {
      navigateToSignUpConfirm,
      navigateToSignUp,
      navigateToForgotPassword,
      handleSubmit,
      handleUsernameChange,
      handlePasswordChange,
      setPasswordInputRef,
      state: { username, password, error },
      props: { getMeErrorMessage, isLoginInFlight: isInFlight, loginError },
    } = this;

    return (
      <FormScene>
        <Text testID="login-title" style={[scene.titleCopy, scene.textCenter]}>
          Login
        </Text>
        <Input
          label="Email"
          onChangeText={handleUsernameChange}
          value={username}
          placeholder="jane.doe@email.com"
          error={error === 'username' ? 'Please provide a valid email' : ''}
          inputProps={{
            autoCapitalize: 'none',
            editable: !isInFlight,
            spellCheck: false,
            keyboardType: 'email-address',
            returnKeyType: 'next',
            onSubmitEditing: focusPassword,
            testID: 'login-email-input',
          }}
        />
        <Input
          label="Password"
          onChangeText={handlePasswordChange}
          value={password}
          placeholder="••••••••"
          error={error === 'password' ? 'Please provide a password' : ''}
          inputProps={{
            ref: setPasswordInputRef,
            secureTextEntry: true,
            editable: !isInFlight,
            spellCheck: false,
            returnKeyType: 'go',
            onSubmitEditing: handleSubmit,
            testID: 'login-password-input',
          }}
        />
        {loginError.length > 0 && <Well text={this.getFormattedLoginError()} />}
        {getMeErrorMessage.length > 0 && (
          <Well text={`Error retrieving user data: ${getMeErrorMessage}`} />
        )}
        <View style={forms.buttonRow}>
          <View style={styles.submitContainer}>
            <Button
              onPress={handleSubmit}
              title="Submit"
              isInFlight={isInFlight}
              testID="login-submit"
            />
          </View>
        </View>
        <View>
          <TouchableOpacity
            accessibilityLabel="Forgot your password"
            onPress={navigateToForgotPassword}
            style={styles.forgotPasswordButton}
          >
            <Text style={[scene.bodyCopy, styles.forgotPasswordText]}>
              Forgot your password?
            </Text>
          </TouchableOpacity>
        </View>
        <View style={styles.noAccountTextContainer}>
          <Text style={scene.bodyCopy}>{"Don't have an account?"}</Text>
        </View>
        <View style={[forms.buttonRow, styles.noAccountButtonRow]}>
          <View style={styles.confirmCodeWrapper}>
            <Button
              onPress={navigateToSignUpConfirm}
              buttonStyles={BUTTON_STYLES.SECONDARY_SKELETON}
              title="Confirm Code"
            />
          </View>
          <View style={styles.signUpWrapper}>
            <Button
              testID="login-signup"
              onPress={navigateToSignUp}
              buttonStyles={BUTTON_STYLES.INFO_SKELETON}
              title="Sign Up"
            />
          </View>
        </View>
      </FormScene>
    );
  }
}

export default connect(
  (state) => ({
    relationshipId: state.relationship.id,
    loverRequestId: state.loverRequest.id,
    getMeErrorMessage: state.user.getMeErrorMessage,
    isLoginInFlight: state.user.isLoginInFlight,
    loginError: state.user.loginError,
  }),
  {
    login: loginAction,
  },
)(Login);
