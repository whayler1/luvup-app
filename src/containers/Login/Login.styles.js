import { StyleSheet } from 'react-native';

import { vars } from '../../styles';

const styles = StyleSheet.create({
  title: {
    fontSize: 30,
  },
  label: {
    marginTop: 20,
  },
  error: {
    color: 'red',
  },
  submitText: {
    textAlign: 'center',
  },
  submitContainer: {
    flex: 1,
  },
  wellError: { marginTop: 32, marginBottom: 0 },
  noAccountTextContainer: {
    marginTop: 32,
  },
  noAccountButtonRow: { marginTop: 16 },
  confirmCodeWrapper: {
    flex: 0.5,
    paddingRight: 8,
  },
  signUpWrapper: {
    flex: 0.5,
    paddingLeft: 8,
  },
  forgotPasswordButton: {
    marginTop: 12,
  },
  forgotPasswordText: {
    textAlign: 'right',
    // fontFamily: vars.fontRegular,
    // fontSize: 18,
    // color: vars.blueGrey500,
  },
});

export default styles;
