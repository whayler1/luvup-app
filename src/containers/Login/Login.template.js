import React from 'react';
import { View, Text, TextInput, TouchableOpacity } from 'react-native';
import { Button } from 'react-native-elements';

import styles from './Login.styles';
import { scene, forms, buttons, modal, wells, vars } from '../../styles';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

export default ({
  navigateToSignUpConfirm,
  navigateToSignUp,
  navigateToForgotPassword,
  onSubmit,
  onUsernameFocus,
  onPasswordFocus,
  onUsernameChange,
  onPasswordChange,
  onBlur,
  username,
  password,
  error,
  isInFlight,
  focus,
  getMeErrorMessage,
}) => {
  /**
   * JW: This is probably not the best way to set input focus on refs,
   * but works for now™
   */
  let passwordInput;
  const focusPassword = () => passwordInput.focus();

  return (
    <KeyboardAwareScrollView>
      <View testID="login-scroll-view" style={scene.content}>
        <Text testID="login-title" style={modal.title}>
          Login
        </Text>
        <View style={forms.formGroup}>
          <Text style={forms.label}>Email</Text>
          <TextInput
            style={[forms.input, focus === 'username' && forms.inputFocus]}
            onFocus={onUsernameFocus}
            onBlur={onBlur}
            onChangeText={onUsernameChange}
            value={username}
            maxLength={50}
            placeholder="jane.doe@email.com"
            placeholderTextColor={vars.blueGrey100}
            autoCapitalize={'none'}
            editable={!isInFlight}
            spellCheck={false}
            keyboardType="email-address"
            returnKeyType="next"
            onSubmitEditing={focusPassword}
            testID="login-email-input"
          />
          {error === 'username' && (
            <Text style={forms.error}>Please provide a valid email</Text>
          )}
        </View>
        <View style={forms.formGroup}>
          <Text style={forms.label}>Password</Text>
          <TextInput
            ref={el => (passwordInput = el)}
            style={[forms.input, focus === 'password' && forms.inputFocus]}
            onFocus={onPasswordFocus}
            onBlur={onBlur}
            onChangeText={onPasswordChange}
            value={password}
            secureTextEntry
            maxLength={50}
            placeholder="••••••••"
            placeholderTextColor={vars.blueGrey100}
            editable={!isInFlight}
            spellCheck={false}
            returnKeyType="go"
            onSubmitEditing={onSubmit}
            testID="login-password-input"
          />
          {error === 'password' && (
            <Text style={styles.error}>Please provide a password</Text>
          )}
        </View>
        {(error === 'credentials' || error === 'server') && (
          <View style={[wells.error, styles.wellError]}>
            <Text style={wells.errorText}>Invalid email or password</Text>
          </View>
        )}
        {getMeErrorMessage.length > 0 && (
          <View style={[wells.error, styles.wellError]}>
            <Text style={wells.errorText}>{getMeErrorMessage}</Text>
          </View>
        )}
        <View style={forms.buttonRow}>
          <View style={styles.submitContainer}>
            <Button
              onPress={onSubmit}
              containerViewStyle={buttons.container}
              buttonStyle={buttons.infoButton}
              textStyle={buttons.infoText}
              title={isInFlight ? 'Submitting…' : 'Submit'}
              disabled={isInFlight}
              testID="login-submit"
            />
          </View>
        </View>
        <View>
          <TouchableOpacity
            accessibilityLabel="Forgot your password"
            onPress={navigateToForgotPassword}
            style={styles.forgotPasswordButton}>
            <Text style={styles.forgotPasswordText}>Forgot your password?</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.noAccountTextContainer}>
          <Text style={scene.copy}>{"Don't have an account?"}</Text>
        </View>
        <View style={[forms.buttonRow, styles.noAccountButtonRow]}>
          <View style={styles.confirmCodeWrapper}>
            <Button
              onPress={navigateToSignUpConfirm}
              containerViewStyle={buttons.container}
              buttonStyle={buttons.secondarySkeletonButton}
              textStyle={buttons.secondarySkeletonText}
              title={'Confirm Code'}
            />
          </View>
          <View style={styles.signUpWrapper}>
            <Button
              testID="login-signup"
              onPress={navigateToSignUp}
              containerViewStyle={buttons.container}
              buttonStyle={buttons.infoSkeletonButton}
              textStyle={buttons.infoSkeletonText}
              title={'Sign Up'}
            />
          </View>
        </View>
      </View>
    </KeyboardAwareScrollView>
  );
};
