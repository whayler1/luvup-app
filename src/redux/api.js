import superagent from 'superagent';

import config from '../config';

const TIMEOUT_DEFAULTS = {
  response: 10000,
  deadline: 30000,
};

const api = () => ({
  post: (path = '', ...args) =>
    superagent
      .post(`${config.baseUrl}${path}`, ...args)
      .timeout(TIMEOUT_DEFAULTS),
});

export const graphqlQuery = query =>
  superagent.post(config.graphQlUrl, { query }).timeout(TIMEOUT_DEFAULTS);

export default api;
