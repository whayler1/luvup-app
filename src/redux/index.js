import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import user from './user/user.reducer';
import relationship from './relationship/relationship.reducer';
import lover from './lover/lover.reducer';
import loverRequest from './loverRequest/loverRequest.reducer';
import coin from './coin/coin.reducer';
import jalapeno from './jalapeno/jalapeno.reducer';
import font from './font/font.reducer';
import userEvents from './userEvents/userEvents.reducer';

export const reducer = combineReducers({
  user,
  relationship,
  lover,
  loverRequest,
  coin,
  jalapeno,
  font,
  userEvents,
});

export const store = createStore(
  reducer,
  applyMiddleware(thunk)
);
