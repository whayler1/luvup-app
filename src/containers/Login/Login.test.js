import React from 'react';
import Login from './Login';

import renderer from 'react-test-renderer';
import { store } from '../../redux';

it('renders without crashing', () => {
  const rendered = renderer.create(<Login store={store} />).toJSON();
  expect(rendered).toBeTruthy();
  expect(rendered).toMatchSnapshot();
});
