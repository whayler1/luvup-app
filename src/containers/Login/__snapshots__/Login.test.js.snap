// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
/>
`;
